sql/src/main/java/org/apache/druid/sql/SqlExecutionReporter.java

```diff
@@ -0,0 +1,146 @@
+
+  private final AbstractStatement stmt;
+  private final String remoteAddress;
+  private final long startMs;
+  private final long startNs;
+  private Throwable e;
+  private long bytesWritten;
+
+  public SqlExecutionReporter(
+      final AbstractStatement stmt,
+      final String remoteAddress
+  )
+  {
+    this.stmt = stmt;
+    this.remoteAddress = remoteAddress;
+    this.startMs = System.currentTimeMillis();
+    this.startNs = System.nanoTime();
+  }
+
+  public void failed(Throwable e)
+  {
+    this.e = e;
+  }
+
+  public void succeeded(final long bytesWritten)
+  {
+    this.bytesWritten = bytesWritten;
+  }
+
+  public void emit()
+  {
+    final boolean success = e == null;
+    final long queryTimeNs = System.nanoTime() - startNs;
+
+    ServiceEmitter emitter = stmt.sqlToolbox.emitter;
+    PlannerContext plannerContext = stmt.plannerContext;
+    try {
+      ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
+      if (plannerContext != null) {
+        metricBuilder.setDimension("id", plannerContext.getSqlQueryId());
+        metricBuilder.setDimension("nativeQueryIds", plannerContext.getNativeQueryIds().toString());
+      }
+      if (stmt.resourceActions != null) {
+        metricBuilder.setDimension(
+            "dataSource",
+            stmt.resourceActions
+                            .stream()
+                            .map(action -> action.getResource().getName())
+                            .collect(Collectors.toList())
+                            .toString()
+        );
+      }
+      metricBuilder.setDimension("remoteAddress", StringUtils.nullToEmptyNonDruidDataString(remoteAddress));
+      metricBuilder.setDimension("success", String.valueOf(success));
+      emitter.emit(metricBuilder.build("sqlQuery/time", TimeUnit.NANOSECONDS.toMillis(queryTimeNs)));
+      if (bytesWritten >= 0) {
+        emitter.emit(metricBuilder.build("sqlQuery/bytes", bytesWritten));
+      }
+
+      final Map<String, Object> statsMap = new LinkedHashMap<>();
+      statsMap.put("sqlQuery/time", TimeUnit.NANOSECONDS.toMillis(queryTimeNs));
+      statsMap.put("sqlQuery/bytes", bytesWritten);
+      statsMap.put("success", success);
+      QueryContext queryContext;
+      if (plannerContext == null) {
+        queryContext = stmt.queryPlus.context();
+      } else {
+        statsMap.put("identity", plannerContext.getAuthenticationResult().getIdentity());
+        queryContext = stmt.queryPlus.context();
+        queryContext.addSystemParam("nativeQueryIds", plannerContext.getNativeQueryIds().toString());
+      }
+      final Map<String, Object> context = queryContext.getMergedParams();
+      statsMap.put("context", context);
+      if (e != null) {
+        statsMap.put("exception", e.toString());
+
+        if (e instanceof QueryInterruptedException || e instanceof QueryTimeoutException) {
+          statsMap.put("interrupted", true);
+          statsMap.put("reason", e.toString());
```

Comment:
I am unfamiliar with reporting the metrics, but is there an advantage of duplicating `exception` as `reason` in case the query is interrupted? Also, in case the query is not interrupted, then should we have `statsMap.put("interrupted", false);` for symmetry?

Reply:
To be honest, I didn't pay too much attention to this code. If you were to diff this function against the original copy, you'd see that only the field names changed: the rest of the code is pretty much how I found it.

That said, yes, it does appear that this code evolved over time and is perhaps not the most optimally structured.

BTW: this is one big hole in the otherwise great GitHub diff mechanism: no easy way to compare code when it moves from one file to another.